### ISSUE TITLE
More examples how to work with JSON columns in postgres

### ISSUE THREAD

**Issue by cloudever** [mikro-orm/mikro-orm#1261]
Is it possible to illustrate more examples how to work with JSON columns in postgres including selects and joins? Thanks

**Comment by B4nan** [comment]
there is no special treatment, so better to read the postgres docs :] if you want to join by anything else than a PK of entity, you will need to use QB (or even raw query).

here is how you can query by custom sql fragment (so possibly a json expression):

https://mikro-orm.io/docs/entity-manager#using-custom-sql-fragments

you could also use `@Formula` decorator to define expression property that could extract something from a json column

https://mikro-orm.io/docs/defining-entities/#formulas

only thing that uses json columns under the hood are embeddables in object mode:

https://mikro-orm.io/docs/embeddables/#storing-embeddables-as-objects

if you have something specific on your mind, feel free to ask

**Comment by aryraditya** [comment]
I tried native json operatorwith postgress driver got some error

```typescript
test('Test JSON', async (t) => {
  const app = await build(t);

  const r = await app.ORM.em.find(TestEntity, {
    [expr('(column1->>"verified")::boolean')]: true
  });

  console.log({ r });
});
```

![image](https://user-images.githubusercontent.com/8693368/103730845-bc1a1900-501e-11eb-869a-fe12a44c4cd9.png)


do mikro-orm support native json operator ?

**Comment by B4nan** [comment]
I see, this is not working as expected as the purpose was mainly to allow using SQL functions. As a workaround, you could do this:

```ts
  const r = await app.ORM.em.find(TestEntity, {
    [expr('jsonb_extract_path_text(column1, \'verified\')::boolean')]: true,
  });
```

**Comment by evenfrost** [comment]
What if the field has an index? Using `jsonb_extract_path_text` instead of `->>` operator won't utilize the index, slowing down the query pretty badly for large databases.

**Comment by B4nan** [comment]
Custom expressions should be fixed in `4.3.5-dev.67` (once built).

Few examples from the test case:

```ts
const r = await orm.em.find(User, {
  [expr('(address4->>\'street\')::text != \'\'')]: [],
  [expr('lower((address4->>\'city\')::text) = ?')]: ['prague'],
  [expr('(address4->>?)::text = ?')]: ['city', 'Prague'],
  [expr('(address4->>?)::text')]: ['postalCode', '12000'],
});
expect(mock.mock.calls[0][0]).toMatch('select "e0".* ' +
  'from "user" as "e0" ' +
  'where (address4->>\'street\')::text != \'\' and ' +
  'lower((address4->>\'city\')::text) = \'prague\' and ' +
  '(address4->>\'city\')::text = \'Prague\' and ' +
  '(address4->>\'postalCode\')::text = \'12000\'');
```

**Comment by B4nan** [comment]
Check #1384, it will add native support for querying by JSON properties, including type casting in postgres (for number and boolean).
